### ISSUE TITLE
Improve security of token system

### ISSUE THREAD

**Issue by anderspitman** [boringproxy/boringproxy#141]
At the very least client tokens shouldn't have full access to the user's account. See the discussion here:

https://forum.indiebits.io/t/potential-security-issue-with-frontend-token-autentication/61

**Comment by alexsanderp** [comment]
Hi @anderspitman!

First I would like to say that I loved this project! ❤️❤️❤️❤️❤️

 I'm using it to configure TCP tunnels on some clients.

I have the same problem, where the token used for the tunnels has access to the web interface, where client can make changes to the registered tunnels.

I would like to know if the idea would be to separate the tokens? Having a token for tunnels and another for web interface?

**Comment by anderspitman** [comment]
Hey @alexsanderp, glad it's useful for you. Yeah basically the idea is to either have 2 completely different types of tokens, or scope the tokens in some way at creation time so you can limit their capabilities. For example, you could limit a token to a specific client or even subdomain. That's probably what I'll try first, and it actually might not be too difficult.

**Comment by anderspitman** [comment]
@WGrobler @alexsanderp I have a branch which adds basic token scoping. The way I implemented it, when you create a token from the web UI, you can choose to scope it to a specific client. If you do, the token will only be usable for operations for that client, which include creating the client and getting tunnels for it. It can't be used for the web UI or any other API functions.

Does this seem like a reasonable approach?

You can try the branch here:

https://github.com/boringproxy/boringproxy/tree/limit-tokens

Pre-built executables here:

https://gemdrive.io/apps/delver/?drive=https://files.apitman.com&path=/public/boringproxy/limit-tokens

**Comment by WGrobler** [comment]
@alexsanderp, I had a look at the new server you created and have a few questions.

I was able to log into the web UI using the admin token, but if I create a token with no limits for a user, the user is unable to log into the web UI with the token, receiving the error "This token is limited to a specific client and cannot be used for the web UI". Is your idea to have only admins access the web UI or am I doing something wrong and regular users can also obtain a fully privileged token.

Generating a new token gives the option to limit to a client, but a client needs to be connected to be part of this list. Is the idea that a new client is set up with the fully privileged token and then a new (limited) token is generated and the full access token is replaced on the client side?

I don't know if its necessary to limit tokens to a single client. I understand that the tokens are used for API calls including client registration, so it makes sense to limit the token to a single client, but that removes the option for automatic deployment. 
Lets say we want to set up a raspberry pi image that automatically connects to your server. If this image includes a token to register a new client, any new instance of the image will automatically connect to the server and the user with its fully privileged token can create the tunnel for the client. Granted, in this example the client name needs to be dynamically changed for each new instance, but that can be done with a local script. The issue with using a single token for automatic creation, is that revoking the token, if abused, will revoke all clients using it. In this case your approach is better. 

Another options is to make tokens limited  to a single client and set the 'client name' when creating the token? This will solve the issue of the client not being part of the dropdown list when a new token is generated. It will also allow the user to change the client name from the web UI and revoke tokens/clients. The user/token combination is used to authenticate the client, so I don't really see any reason to set the client name on the client side.  Was there any reason for this other than identifying different clients using the same token.
We can then create some sort of automatic deployment token that can be used to call the API and request a new client token. In my Pi example, a script can be set up to call the server API using the deployment token and the server can provide a client token for tunnelling. If the deployment token gets revoked, no new clients can join using the deployment token, but existing client can continue using their client tokens since the deployment token was only used for initial authentication.

**Comment by anderspitman** [comment]
Sounds like this is a combination of a couple bugs in my implementation, and some use cases I hadn't considered. I don't have time right at the moment to take another look, but I agree with pretty much everything you're saying and I'll get back to you soon.

**Comment by anderspitman** [comment]
Ok I've thought a bit more about this. It's a bit complicated because I'm also in the process of designing the OAuth2 protocol, which is closely related.

I think the key change is that client names need to be created in the web UI, either manually (perhaps when creating tokens) or during the OAuth2 flow.

I'm not sure I understand how to properly implement things to handle your deployment case. Are you saying you already have a way to handle this, or do we need additional changes?

**Comment by anderspitman** [comment]
The broken login token bug for non-admin users should be fixed. It actually wasn't even working for admin for me.

Current plan is to add a "Clients" page to the UI designed similarly to the Tokens page, and require Clients to be added manually before use.

**Comment by anderspitman** [comment]
Alright current status is I've implemented Clients page, and also simplified the API so the user and client-name are no longer needed in every case when running the client, since that information can often be obtained from the token used for the request (see #66).

This is ready for testing whenever you get a chance @WGrobler:

https://gemdrive.io/apps/delver/?drive=https://files.apitman.com&path=/public/boringproxy/limit-tokens

**Comment by WGrobler** [comment]
>This is ready for testing whenever you get a chance @WGrobler:

Thanks @anderspitman, I will test today and give feedback

> I'm not sure I understand how to properly implement things to handle your deployment case. Are you saying you already have a way to handle this, or do we need additional changes?

This is a hypothetical use case. I haven't implemented anything at this time. When I opened the original discussion, I wanted to start discussions around changed to ensure that the authentication system is secure but also futureproof, since the project is in beta its a good time to make core changes. I don't think its required to add automatic deployment tokens at this time, but I think its useful to keep in mind. 

Since you already made changes to have multiple typed of tokens, this forms the basis for adding more types of tokens in the future. When I have time, I'll draw up the logic for something like this.

**Comment by WGrobler** [comment]
> This is ready for testing whenever you get a chance @WGrobler:
>https://gemdrive.io/apps/delver/?drive=https://files.apitman.com&path=/public/boringproxy/limit-tokens

@anderspitman I tried the version you supplied, but keeps getting an error "This token is limited to a specific client and cannot be used for the web UI" when trying to access using the admin token. I deleted the DB to start fresh and it still won't give me access to the web UI

**Comment by anderspitman** [comment]
It's possible I accidentally linked the wrong files. I just rebuilt and uploaded to be sure.

**Comment by WGrobler** [comment]
I found the issue,

In [ui_handler.go](https://github.com/boringproxy/boringproxy/blob/limit-tokens/ui_handler.go):92 you check if tokenData.Client is not "", but in [boringproxy.go](https://github.com/boringproxy/boringproxy/blob/limit-tokens/boringproxy.go):136 you create a new admin with `db.AddToken("admin", "any")`.

Either the if statement need to change to look for "any" instead of no-string or the Add user statement needs to change. 

Since previous versions did not create a client entry for admin, perhaps the check should be for not null and not "any"?

**Comment by WGrobler** [comment]
By making changes to my boringproxy_db.json to correspond with [ui_handler.go](https://github.com/boringproxy/boringproxy/blob/limit-tokens/ui_handler.go):92, I managed to test the setup.

Its working great, thanks for setting this up. It's a massive security improvement. 

The other improvement that resulted from this, is that I an now add a tunnel before even setting up the client. Since the client is added on the web UI, its already available in the add tunnel dropdown before the client connects for the first time. This is also a massive UX improvement. 

Small required fix: The client tokens under /tokens in the web UI, still shows the login link. This can be removed for tokens without web UI access.

**Comment by anderspitman** [comment]
Oh that makes sense! I didn't even test making a fresh DB. This would be a good argument for some automated testing.

Ok I changed that code to create the initial admin token, and it seems to be working for me for a fresh DB. I also removed the login links for client tokens.

Everything is merged into master. I think between this and WebSockets support it's worth cutting a new release.

Thanks everyone for your feedback and @WGrobler for the suggestion and debugging!
